src/parlant/core/sessions.py

```diff
@@ -677,24 +683,27 @@ async def create_event(
         if not await self._session_collection.find_one(filters={"id": {"$eq": session_id}}):
             raise ItemNotFoundError(item_id=UniqueId(session_id), message="Session not found")
 
-        session_events = await self.list_events(
-            session_id
-        )  # FIXME: we need a more efficient way to do this
-        creation_utc = creation_utc or datetime.now(timezone.utc)
-        offset = len(list(session_events))
+        async with self._lock.writer_lock:
```

Comment:
Why is this a writer lock? And why aren't you locking over `find_one()` as well?

Reply:
Oh wait I understand why it's a writer lock. The UI here gave me the impression it's under `read_session()`. But I think you need to include `find_one()` in the locked scope, otherwise you can assume that it exists and then go on to find nothing.